core/log/solver_debug.cpp

```diff
@@ -0,0 +1,321 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/log/solver_debug.hpp>
+
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/mtx_io.hpp>
+#include <ginkgo/core/base/name_demangling.hpp>
+#include <ginkgo/core/log/logger.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+
+#include "core/base/dispatch_helper.hpp"
+
+
+namespace gko {
+namespace log {
+namespace {
+
+
+template <typename Functor>
+static bool dispatch_type(const LinOp* value, Functor fn)
+{
+    const auto host_exec = value->get_executor()->get_master();
+    using conv_to_double = ConvertibleTo<matrix::Dense<double>>;
+    using conv_to_complex = ConvertibleTo<matrix::Dense<std::complex<double>>>;
+    if (dynamic_cast<const conv_to_double*>(value)) {
+        auto host_vec = matrix::Dense<double>::create(host_exec);
+        host_vec->copy_from(value);
```

Comment:
this is not always a clone, in some cases it can be a conversion. If I didn't use `double`, I would have to handle every float type separately - going through the highest precision type for debugging purposes seemed more appropriate to me than duplicating the amount of code.

Reply:
In that case, maybe it would be better to use `convert_to`?

Reply:
might make things clearer, yes.